File: java/src/solutions/3418_maximum_amount_of_money_robot_can_earn.java
class Solution {
  public int maximumAmount(int[][] coins) {
    final int NEUTRALISES = 2;
    Integer[][] rollingRow = new Integer[coins[0].length][NEUTRALISES + 1];
    rollingRow[0][0] = 0;

    for (int rowIdx = 0; rowIdx < coins.length; rowIdx++) {
      for (int colIdx = 0; colIdx < coins[0].length; colIdx++) {
        int coin = coins[rowIdx][colIdx];
        Integer[] rollingCol = rollingRow[colIdx];

        for (int i = 0; i < NEUTRALISES + 1; i++) {
          Integer left = null;
          if (colIdx != 0) {
            left = rollingRow[colIdx - 1][i];
          }
          Integer top = rollingCol[i];
          if (left == null && top == null) {
            continue;
          } else if (left == null) {
            rollingCol[i] = top + coin;
          } else if (top == null) {
            rollingCol[i] = left + coin;
          } else {
            rollingCol[i] = Integer.max(top, left) + coin;
          }
        }

        if (coin < 0) {
          var snapshot = rollingCol.clone();
          for (int i = 0; i < NEUTRALISES; i++) {
            if (snapshot[i] == null)
              break;
            int neuralisedVal = snapshot[i] - coin;
            if (rollingCol[i + 1] == null || neuralisedVal > rollingCol[i + 1])
              rollingCol[i + 1] = neuralisedVal;
          }
        }
      }
    }

    var last = rollingRow[rollingRow.length - 1];
    int result = last[0];
    for (Integer i : last) {
      if (i == null) {
        continue;
      }
      if (i > result) {
        result = i;
      }
    }
    return result;

  }

  void main() {
    int x = 0;

    x = maximumAmount(new int[][] { { 0, 1, -1 }, { 1, -2, 3 }, { 2, -3, 4 } });
    System.out.println(x);
    assert x == 8;

    x = maximumAmount(new int[][] { { 10, 10, 10 }, { 10, 10, 10 } });
    System.out.println(x);
    assert x == 40;

    x = maximumAmount(new int[][] { { -7, 12, 12, 13 }, { -6, 19, 19, -6 }, { 9, -2, -10, 16 }, { -4, 14, -10, -9 } });
    System.out.println(x);
    assert x == 60;
  }
}

File: rust/src/bin/3418_maximum_amount_of_money_robot_can_earn.rs
struct Solution {}

impl Solution {
    pub fn maximum_amount(coins: Vec<Vec<i32>>) -> i32 {
        let mut dp = vec![[i32::MIN / 2; 3]; coins[0].len() + 1];
        dp[1] = [0, 0, 0];

        for row in coins {
            for (col_idx, coin) in row.iter().enumerate() {
                dp[col_idx + 1][2] = i32::max(
                    i32::max(dp[col_idx + 1][2], dp[col_idx][2]) + coin,
                    i32::max(dp[col_idx + 1][1], dp[col_idx][1]),
                );
                dp[col_idx + 1][1] = i32::max(
                    i32::max(dp[col_idx + 1][1], dp[col_idx][1]) + coin,
                    i32::max(dp[col_idx + 1][0], dp[col_idx][0]),
                );
                dp[col_idx + 1][0] = dp[col_idx + 1][0].max(dp[col_idx][0]) + coin;
            }
        }
        dp.last().unwrap()[2]
    }
}

fn main() {
    let mut x;

    x = Solution::maximum_amount(vec![vec![0, 1, -1], vec![1, -2, 3], vec![2, -3, 4]]);
    println!("{}", x);
    assert_eq!(x, 8);

    x = Solution::maximum_amount(vec![vec![10, 10, 10], vec![10, 10, 10]]);
    println!("{}", x);
    assert_eq!(x, 40);

    x = Solution::maximum_amount(vec![
        vec![-7, 12, 12, 13],
        vec![-6, 19, 19, -6],
        vec![9, -2, -10, 16],
        vec![-4, 14, -10, -9],
    ]);
    println!("{}", x);
    assert_eq!(x, 60);
}
